Initial work

jhford opened the pull request:
Here's the first batch of commits, but it's probably better to take a look at it as the complete diff.  The internal functions are a little bit messy, I need to work on cleaning them up a little bit.  I'd also like to change a lot of parameters to pointers so that I can specify `nil` when calling.

This PR is the underlying code that I need for the special blob artifact methods.  I will make the `createArtifact` and `completeArtifact` calls through the tcclient's Queue.  The request runner here is used to run the requests given back by the `createArtifact` method and downloads done using the artifact's get method.

jhford committed:
Sha256 a file

jhford committed:
Hash a file as either a multipart or a single part file

jhford committed:
Rename module

jhford committed:
Implement a file upload preparer

jhford committed:
Go Format

jhford committed:
Fix remaining issues with the prepare module

jhford committed:
Implementation of a Body type as well as a better set of tests for the code written so far

jhford committed:
Split runner.go into specific files and do go fmt on everything

jhford committed:
better unit testing

jhford committed:
Better error handling in upload preperation

jhford committed:
fix bug in multipart upload preperation

jhford committed:
Fix issues with tests and go format

jhford committed:
Go Format for runner

jhford committed:
Use a constant filename for compressed files

jhford committed:
move prepare into the main module and make it unexported

jhford committed:
Fix a couple small bugs, change exporting

jhford committed:
testing cleanup

jhford committed:
A little more cleanup and test fixing

jhford committed:
Unexport a bunch of stuff

jhford committed:
return errors instead of panic

jhford committed:
More code

jhford committed:
make the bytecountingwriter work

jhford committed:
requests mostly working

jhford committed:
Gzip requests work

jhford committed:
Oops, i forgot to add the requests tests

jhford committed:
cleanup

jhford committed:
go fmt

jhford committed:
fix slight issue with how tests found buffer to run against

jhford committed:
Add a simple logging module

jhford committed:
Go Format

jhford committed:
cleanup

jhford committed:
Start the interface

jhford committed:
Add a ReadSeekCloser interface

jhford commented on prepare.go:
```diff
@@ -0,0 +1,451 @@
+}
+
+// Internal struct for hashing a simple single part file with compression
+type compressedSinglePartFileInfo struct {
```
I was thinking I could probably merge all these internal structs, but I didn't want to need to know the nil value of each member type.

jhford commented:
Still to do:

* implement the full upload/download methods
* write a simple CLI
* possibly move to `ReadSeekCloser` references instead of using filenames

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+package main
```
The package `main` is only for executable code, so we probably want this package to be `artifact` or something similar.

jhford replied:
this might be a misunderstanding of how package naming and imports work in go.  My understanding is that if someone imports `github.com/taskcluster/taskcluster-lib-artifact-go` that they'd get the `main` package, and that if I had something like `package artifact`, then someone would have to import `github.com/taskcluster/taskcluster-lib-artifact-go/artifact`.  Am I misunderstanding that?

petemoore replied:
Yes, when you import `github.com/taskcluster/taskcluster-lib-artifact-go` the master branch of this repo will be checked out to `${GOPATH}/src/github.com/taskcluster/taskcluster-lib-artifact-go`. Then the `*.go` files in that directory will be read, and it will be necessary for them all to have a `package <x>` declaration in them. The value of `<x>` needs to be the same in all the `*.go` files in order for the compile to succeed. The name here can be different to the enclosing directory name, but by convention it is usually the same - note, the package name being the same as the directory name is _not_ enforced. After you have imported this package, the code that uses code in this package will need to refer to exported identifiers as `<x>.<identifier>`. For example, if you name your package with `package artifact` and provide a function `NewBody` in it, then code that calls it will need to look as follows:

```go
package otherpackage

import (
    "github.com/taskcluster/taskcluster-lib-artifact-go"
)

func SomeFunction() {
    body := artifact.NewBody()
}
```

Note, it is tricky to see from this code that `artifact` is the package name of the package located at `"github.com/taskcluster/taskcluster-lib-artifact-go"` which is why by convention we name the package the same as the directory it is located in. With a directory that has hyphens (`taskcluster-lib-artifact-go`) we cannot use the same name, since hyphens are not allowed in package names. In order to make it easier to see the connection between `taskcluster-lib-artifact-go` and `artifact` you can simply name the package at import. This also solves the problem if you import two different packages that clash on package name, e.g.:

```go
package otherpackage

import (
    artifact "github.com/taskcluster/taskcluster-lib-artifact-go"
    artifact2 "github.com/some/package/also/called/artifact"
)

func SomeFunction() {
    body := artifact.NewBody()
    otherArtifact := artifact2.New("apples") // some function in the other artifact package
}
```

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+	"os"
+)
+
+type body struct {
```
For other packages to be able to interact with `body` it will need to be exported (i.e. renamed to `Body`)

jhford replied:
I'm not entirely sure about the visibility.  If I expose a NewBody method that returns body, and I want that returned body to be opaque, do I still need to export it?

I was also thinking that the body struct would be a purely internal function, with only the `Download()` and `Upload()` methods exposed.

petemoore replied:
If calling code needs a handle on the body (even if it is opaque) it will need to be exported, otherwise it won't be able to pass it back to e.g. another function call that receives it, and it won't be possible to call other methods on it. If it is a struct, you can "hide" the individual members by not exporting them. Then anything in your package can use these struct members, but no code outside sees the members or can do anything with them.

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+// Create a body.  A body is an io.Reader instance which reads from the file at
+// filename, starting at the `offset`th byte and reading up to `size` bytes in
+// total.
+func newBody(filename string, offset, size int64) (*body, error) {
```
Likewise, `NewBody` will be needed to be able to call it from a different package.

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+// total.
+func newBody(filename string, offset, size int64) (*body, error) {
+	if size == 0 {
+		return nil, errors.New("Cannot specify a size of 0")
```
Unfortunately errors in go tend to get propagated "up the stack" so sometimes client code might not realise where the error originated from. In this case it might be better to state something like `fmt.Errorf("taskcluster-lib-artifact-go: artifact.NewBody called with size 0 for filename %v and offset %v - size 0 is not allowed.", filename, offset)`.

jhford replied:
:'( I'm sure not a fan of error handling in Go, but that's a good suggestion.

petemoore replied:
Yeah I am also not the greatest fan of error handling in go.

Another (slightly more involved) option is to create a custom type for the error, e.g.:

```go
type NilBodySizeError struct {
    Filename string
    Offset   int64
}

func (n *NilBodySizeError) Error() string {
    return fmt.Sprintf("NilBodySizeError for filename %v and offset %v", n.Filename, n.Offset)
}

// NewBody creates a new artifact body ....
// If size is 0, an error will be returned of type *NilBodySizeError.
func NewBody(filename string, offset, size int64) (body *Body, err error) {
    if size == 0 {
        return nil, &NilBodySizeError{
            Filename: filename,
            Offset: offset,
        }
    }
    ....
}
```

The advantage of this is that calling code sees the error is of type `*artifact.NilBodySizeError` and so knows which package this error comes from, and can also directly see the doc comments for that error type. Also calling code can query the members of the error (`Filename` and `Offset`) and do custom handling based on the values, if it really wanted to.

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+
+	b := body{file, nil, offset, size}
+
+	b.Reset()
```
This could return an error if the file is a directory, so you should check for an error here.

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+	return &b, nil
+}
+
+func (b *body) Reset() error {
```
Some comments for all the functions would be nice.

petemoore commented on body.go:
```diff
@@ -0,0 +1,64 @@
+	return nil
+}
+
+func (b body) Print() {
```
Using `String()` instead of `Print()` would make this method available to the `fmt` package (see [Stringer](https://golang.org/pkg/fmt/#Stringer) interface).

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,166 @@
+	"testing"
+)
+
+var allTheBytes []byte = []byte{1, 3, 7, 15, 31, 63, 127, 255}
```
You could use `const` instead of `var` here.

jhford replied:
When I try:
```go
const allTheBytes []byte = []byte{1, 3, 7, 15, 31, 63, 127, 255}
```
I get:
```
./body_test.go:10: const initializer []byte literal is not a constant
```

I also tried declaring it as a `[8]byte` which also didn't work.  Any idea why it's failing?

petemoore replied:
Thanks John, TIL [arrays can't be constants](https://groups.google.com/forum/#!topic/golang-nuts/5aynucvg96I)! You're quite right. :-)

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,166 @@
+
+	filename := "test-files/body-reading"
+
+	_file, _err := os.Create(filename)
```
I'm curious why these variables start with an `_`? I'm not familiar with this convention.

jhford replied:
I didn't really like having them in the function scope, since they're only initializing code.  Maybe I could instead put that block of code into a function and call it... This was just a cleanliness thing

petemoore replied:
Let's chat about this, I'm not really sure what this means. The scope of the variables shouldn't change according to whether they have a `_` prefix or not, and the variables declared inside a function won't ever be visible to code outside of that function.

jhford replied:
I know that it doesn't change the scope, it was purely a naming thing, I tend to do this because scoping in other lanagues (python, js-var) works a little strangely, so I tend to be careful about giving names that I'm likely to reuse in a nested scope.

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,166 @@
+			t.Error(err)
+		} else if err == nil {
+			t.Error("Expected an error")
+		}
```
I think you could get a nil pointer here if err is nil, maybe more idiomatic would be:

```
if err == nil {
  t.Fatal("Expected an error")
}
if os.IsNotExist(err) {
  t.Fatal(err)
}
```

jhford replied:
i'm not sure how to interpret this, is the concern about using two `if` instead of `else if` or is it that I'm checking a value of `err` before I've verified that it's non-nil?  Out of curiosity, I tested it and `os.IsNotExist(nil)` returns `false` instead of an error.

petemoore replied:
Yes, both of those really. So taking a step back, the main objective of this review was really just to provide some insight into the types of issues you hit in go programming, rather than only being solely about code correctness (which of course is also important). It is quite idiomatic not to chain `if { ... } else if { .... }` in go, but rather to exit in a code block if you are done processing, and then to have a seperate `if` statement afterwards. I remember reading a go blog about this (I'll see if I can find it). Another common practice is to always check for nil first, if you are going need to check for it. In this case it technically isn't needed, it is just a useful habit to get into, and perhaps makes it less likely that bugs slip in when code is refactored to test for other error conditions later, that might require that `err != nil` etc.

jhford replied:
cool!  that's really strange to me about the not chained if-elseif-else... good to know.

I totally agree that a nil check should've been first though!

petemoore replied:
So for reference:

* https://github.com/golang/go/wiki/CodeReviewComments#indent-error-flow
* [golint](https://github.com/golang/lint) reports non-conformance

petemoore commented on interface.go:
```diff
@@ -0,0 +1,25 @@
+func New(creds *tcclient.Credentials) *Artifact {
+  q := queue.New(creds)
+  a := newAgent()
+  return Artifact{q, a}
```
`return &Artifact{q, a}`

petemoore commented on log.go:
```diff
@@ -0,0 +1,26 @@
+// file.  We are doing to ensure that by default, consumers of this library get
+// useful logs for debugging.  We do, however, want to allow the users to
+// change where logging should go, so we'll expose methods to change the
+// behaviour of logging.
```
Nice!

petemoore commented on main.go:
```diff
@@ -0,0 +1,5 @@
+package main
+
+func main() {
```
Obviously this is just a placeholder, but when you do create a main function, you'll probably want to put it in a different package (maybe a subdir called `cmd/<command name>` or something like that) and put that in a main package.

jhford replied:
I'm curious why I would want to put a command into a different package.  Wouldn't it be ideal if I could import the library and use it as a library and the same package when built on its own would be a CLI tool?  I'm probably missing something about how go programs are organized.

petemoore replied:
That would be nice, But unfortunately in that case, all calling could would need to call e.g. `main.NewArtifact()` rather than `artifact.New()` etc. For this reason, we're really forced to put them in different packages.

The thing that might make this easiest to understand is if you create a second project which imports your package and uses it. I think as soon as you start trying to call it from outside the package, which is what your library users will also need to do, it will start to make sense.

jhford replied:
if someone imported the library without giving it a name, e.g. `import ("github.com/taskcluster/taskcluster-lib-artifact-go")`, how would they access the Download() method?  would they have to give it their own name (e.g. `import (artifact "github.com/taskcluster/taskcluster-lib-artifact-go")`) then use artifact?   Are you saying that if I made it `package artifact` that `import ("github.com/taskcluster/taskcluster-lib-artifact-go/artifact")` would allow someone to use `artifact.Download()`.

I think I'm getting this now...  Go modules feel strange to me still

petemoore replied:
Yes that is about right, but with the additional hurdle that when you compile a package called main, it doesn't create a library, it creates a static binary. You can only import libraries, so you wouldn't be able to import something that contained a `main` package. Note, the package represents only one directory level - anything in a subdirectory is a different package - so packages are a little bit different here to other languages. The scope of the package is just the directory it is in.

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,166 @@
+package main
```
Note, since you're testing a library, rather than a command line utility, it would be useful to give the tests a different package name (such as `test`). This will force the tests to refer to the fully qualified name of the go types they interact with (e.g.

```
_, err := artifact.NewBody("file", 128, 128)
```

The advantage of this is that your test will be calling the library in the same way that a user would call it, and you can be sure that your library has exported the types that are needed by the user.

Note when you move it out of `main` into e.g. `artifact` you can also generate the go docs, and you'll see there which functions/methods/variables are visible externally, which is also useful for seeing how your API is presented to users of it.

jhford replied:
My goal is to not have newBody be an exposed function.  It's really just an implementation detail.  Given that, would you still recommend doing it this way?

petemoore replied:
Sorry I had wrongly assumed calling code would call this. Indeed, if `newBody` is only called from code inside the package, i.e. calling code calls other functions/methods and doesn't see `newBody`, then yes, not exporting it would be the right choice (i.e. keeping it named `newBody`).

petemoore reviewed (commented):
It looks great John. The most useful tip I think I can give is that it would be really useful to move everything out of the main package, and maybe even split into several smaller packages, but having tests call into the library from a different package. Note they can still live in the same directory, just have a different package name, so the code and tests will still be together in same directory, just in a different package so that visibility laws will be applied. I think once you do that, it will be easy to see what surface your APIs expose, and what should be internal/external, and at that point it might make it easier to also break the library into smaller packages if necessary.

I focussed mostly on code organisation in this review, I'm sorry about that, once that is done though, I'd be happy to dig in more in-depth, I didn't want to overload the first review.

jhford committed:
Address some review feedback

jhford committed:
Move setup code into a function

jhford committed:
use Fatal instead of Error for test failures

jhford committed:
More tests and better stuff

jhford committed:
Address some lint

jhford committed:
Some more code

jhford committed:
Address some feedback and use io.Writer instead of a filename for output

jhford committed:
Test that writing the response body works as expected

jhford committed:
Allow passing in a nil body

jhford committed:
Ugh, typo

jhford committed:
Make downloads work and fix remaining issues on uploads

jhford committed:
hack until taskcluster/taskcluster-queue #220 is deployed

jhford committed:
create body instances based on a ReadSeekCloser instead of Files

jhford committed:
Convert to using io interfaces instead of strings and simplify upload preperation

jhford committed:
Fix up tests

jhford committed:
Simplify and fix the upload preperation tests

jhford committed:
We really don't need ReadSeekClosers

jhford committed:
Fix up and improve the interface tests.  There's still a failure that I'm tracking down

jhford committed:
Fix bug where we were reading from the input when we should be reading from output

jhford committed:
Clean up interface and start some benchmarks

jhford committed:
remove some benchmarks that are so slow we can ignore them

jhford commented:
Hey @petemoore it'd be great if you could give this another look.  I might push some commits until I start seeing review comments, but once that happens I'll start a new branch.

A couple of notes:

1. <strike>I have to force 3bed0de6f75d4cd38e2309c5b7fc02b2ef2d6c2a for taskcluster-client-go because of the int vs. int64 issue.</strike> not needed?
2. You'll need valid TASKCLUSTER_{CLIENT_ID,ACCESS_TOKEN} environment variables to run the integration tests
3. I have a simple benchmark for preparation, but I'm not sure if I've structured it correctly.  I think what's happening is that the BenchmarkPrepare is overall taking > the timeout (10m), so it's killing the whole thing.  I'm not sure how to fix this behaviour.  Basically, this is to check what chunkSize is ideal.  That value determines how many bytes to read at any given time.
4. This code is drastically cleaner, and with fewer types than the first PR.  I think embracing nil values really enabled this to be a lot tidier.  I went from 2 single part file info structs, 1 multipart file info struct, 1 single part upload info struct, and 1 multi part upload info struct down to a singlle "upload" struct which is basically built up as the calls happen.  This drastically simplified the code
5. I'm no longer opening files directly, but rather taking as arguments the minimum io interface needed for the operations.  I'm really unsure of what benefit using a ReadSeekCloser interface would be over a ReadSeeker, if I'm not closing the interface, so I've stuck to upstream interfaces.  This means that the caller is going to be responsible for creating the streams, including the scratch file for uploads.  I think this is a fair design because the library user will best know where tmp files ought to be placed.  Here's what I'm talking about:
```go
// Ignoring error handling for brevity
input, _ := os.Open("my-firefox-build.zip")
defer input.Close()
tmpfile, _ := ioutil.TempFile("/tmp", "artifact-upload-scratch")
defer tmpfile.Close()
defer os.Remove(tmpfile.Name())
client.Upload("tid", "rid", "name", input, tmpfile, true, false)
```
6. In the body.go:body struct, I have a .Close() method, but the underlying reader doesn't need to be a Closer.  I'm using a type assertion to say that if you call the body.Close() method, and your underlying reader happens to also be a closer, that you should call .Close() on the underlying reader.  I'm not sure if this is a good idea
7. I've implemented content-type sniffing using go's built in sniffer which follows the whatwg spec, aiui.  I'm curious whether you think this is a good idea, or whether you think we should just add a param for content-type to the Upload() method.

jhford committed:
Benchmarking showed 128KB chunksize as being faster

jhford committed:
tidy up integration test logs as well

jhford committed:
A little more validation on the redirects

jhford committed:
Use the hex package to do hex encoding

jhford committed:
Move interface errors to an errors.go file

jhford committed:
Better docs!

jhford committed:
A Script for running in CI

jhford committed:
Cleanup and lint

jhford committed:
Cleanup and lint stuff

jhford committed:
Better headers

jhford committed:
Hmmmmm

jhford committed:
Better godoc to markdown

jhford committed:
better readme

jhford committed:
CLI Tool

jhford committed:
cleanup and hack removal

jhford committed:
Cleanup interface tests a little with a closure

jhford committed:
cleanup body_test.go

jhford committed:
unit tests for the CLI

jhford committed:
Some tests for the CLI client

jhford committed:
Fix the CLI tests and the CLI

petemoore commented on test-files/readme:
```diff
@@ -0,0 +1,4 @@
+This directory is used to store scratch files for the unit tests of this
+library.  Nothing important is stored in here.  We do this instead of using
+/tmp to make sure that the /tmp mount doesn't get filled up with a bunch of big
+files
```
See https://golang.org/cmd/go/#hdr-Test_packages - `testdata` would be the idiomatic name for this directory instead of `test-files`.

> The go tool will ignore a directory named "testdata", making it available to hold ancillary data needed by the tests.

jhford replied:
sounds good!

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+const (
+	ErrInternal = 1
+	ErrBadUsage = 2
+	ErrCorrupt  = 13
```
`man sysexits` might offer some alternative suggestions for values here - although not required, might be nice.

jhford replied:
good thought!  I can't seem to load that manpage on Fedora, but the header is easy enough to understand.  0fcffd0

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+)
+
+// TODO implement an in memory 'file'
+// TODO implement 'redirect' and 'error' artifact types?
```
Should these TODOs be in the library package?

jhford replied:
yep, they probably should be.  0a226bc

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+
+func _main(args []string) error {
+	// We're going to take care of exiting ourselves
+	cli.OsExiter = func(c int) {}
```
Too bad you don't like `docopt` 😜

jhford replied:
I originally tried it, actually.  I found the lack of "real" subcommand support off-putting, but the good news is that this library tackles the drift between docs and reality by generating really nice help messages.

petemoore replied:
fwiw now i've seen this library, i like it a lot too - it is quite intuitive :-)

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+		},
+		cli.IntFlag{
+			Name:   "chunk-size",
+			Usage:  "set the I/O chunk size to `CHUNK_SIZE` KB",
```
Really this should say [KiB](https://physics.nist.gov/cuu/Units/binary.html) according to SI units.

An alternative would be to use `cli.Int64Flag` and specify all content sizes in bytes. This would help with consistency across the command line parameters. It also is in keeping with e.g. `Content-Length` header being in bytes. Another alternative might be to allow the parameter to specify units (e.g. `--chunk-size 5 MiB`).

petemoore replied:
FWIW I had a play, and this looks like a useful library for this purpose:

```go
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ cat main.go 
package main

import (
	"fmt"
	"log"
	"strings"

	"github.com/alecthomas/units"
)

func main() {
	const size = "24.3 MiB"
	bytes, err := units.ParseBase2Bytes(strings.Replace(size, " ", "", -1))
	if err != nil {
		log.Fatalf("Woooah: %v", err)
	}
	fmt.Printf("%v == %#v bytes\n", size, bytes)
}
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ go run main.go 
24.3 MiB == 25480396 bytes
pmoore@Petes-iMac:~/go/src/github.com/taskcluster/testparse $ 
```

jhford replied:
There's some debate about whether KiB or KB, as opposed to kB is the best option.  I personally find KiB ugly, so tend to use KB, plus the only people who really care about 1000 byte increments are those short changing us on physical storage :(

Regarding the chunk size, it would be in line with Content-Length, but based on benchmarks, anything lower than about 32KB is so slow it'd be useless, which means that we'd be having to do things like `--chunk-size 65536` instead of `--chunk-size 64`.

Similarly, for multi-part, the S3 enforced sizes are 5MB-5GB, which means that we'd always be passing 5242880-5368709120 instead of 5-5000.

Both of these options are not intended to be used commonly, but I think the pragmatism of using the KB and MB is worthwhile.  Alternately, do you know a good go library which does parsing of "1 KB" into a number 1024?  That'd be a nicer solution all around!

petemoore replied:
See code sample above - I really like github.com/alecthomas/units. Funnily enough he is also the author of gometalinter and also the python voluptuous library.

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+					Usage: "serve artifact with gzip content-encoding",
+				},
+				cli.BoolFlag{
+					Name:  "multi-part",
```
The word [multipart](https://www.w3.org/Protocols/rfc1341/7_2_Multipart.html) isn't hyphenated so this should be `multipart` rather than `multi-part`.

It is one of those strange cases like backslash vs forward-slash, where one is hyphenated and the other isn't. That's the English language for you! 😄

jhford replied:
haha, i've been using hyphenated multi-part throughout this entire project!  I wonder how big the diff to fix this in the go library would be.

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+		Certificate: os.Getenv("TASKCLUSTER_CERTIFICATE"),
+	}
+
+	q, err := queue.New(creds)
```
You can do this using:

```go
q, err := queue.New(nil)
```

and don't need to create `creds` variable.

jhford replied:
I thought I removed all of those, but i missed this one, thanks for spotting it.

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+	tcr := queue.TaskClaimRequest{WorkerGroup: "my-worker-group", WorkerID: "my-worker"}
+	tcres, err := q.ClaimTask(taskID, "0", &tcr)
+
+	tEnv.runID = strconv.FormatInt(tcres.RunID, 10)
```
note: you can also use `strconv.Atoi(tcres.RunID)` although maybe it is easier to guess what `FormatInt` is than `Atoi`. 🤣

jhford replied:
So the issue is that the tcclient types now return int64 as the runID (which is the desired behaviour), but `Itoa` only takes 32-bit ints, and there isn't an `Itoa64`.  From the docs: ` Itoa is shorthand for FormatInt(int64(i), 10).`.  Good thought though, it's what i'd originally used.

petemoore replied:
Fair enough!

petemoore commented on errors.go:
```diff
@@ -0,0 +1,25 @@
+
+// ErrBadOutputWriter is returned when the output writer passed into a function
+// was able to be checked for its size and it contained more than 0 bytes
+var ErrBadOutputWriter = errors.New("output writer is not empty")
```
In general I feel there is a lot more value in an error that says "Resource xyz is corrupt" or "Unexpected redirect from url xyz" than the more general error "Corrupt resource" or "Unexpected redirect". When seen in logging, it enables you to troubleshoot much more effectively. For this reason I would prefer we have custom error types that include the instance data that caused the failure (the url that had the unexpected redirect, etc) and they can have a `func Error() string` method that returns a descriptive error containing that pertinent data. It is more work upfront, but saves a lot of time and work later. The type of error can still be detected in a [type switch](https://tour.golang.org/methods/16) statement.

jhford replied:
`error.go` has a lot of work now which produces errors like this:

```
Artifact Error:
  1. (internal) downloading http://127.0.0.1:40011/the-redirect to write to test-file-output728894125
  2. (internal) corrupt resource
```

petemoore reviewed (changes requested):
Hey John,

I haven't totally finished the review, but I thought it was worth sending this over already, so you have something to start on, and I'll carry on looking tomorrow!

Thanks, and apologies if any of my suggestions cause any strong feelings of disagreement, we can talk about anything and everything, so don't feel like everything I write is black-and-white! And I'm in IRC as always if you want to discuss any of it. :-)

jhford committed:
Use exit codes inspired by sysexits.h

jhford committed:
Move notes about library features into library

jhford committed:
Multipart is a single word

find * -name "*.go" -type f -exec sed -i -e "s/multi-part/multipart/g ; s/MultiPart/Multipart/g ; s/multiPart/multipart/g" {} \;

jhford committed:
Use tcclient's credential loader

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+				}
+
+				if !c.IsSet("output") {
+					return cli.NewExitError("must specify output", ErrBadUsage)
```
Shouldn't this write to standard out instead?

> `If not provided, Standard Output will be assumed`

jhford replied:
I've changed it so that you have to use "-" explicitly like many other unix utilities.  I suspect that's nicer than getting whole artifacts accidentally printed to logs by forgetting the output parameter

jhford replied:
actually, stdout doesn't support seeking even though it is a `os.File*`, which causes things to go badly.  Since I'm using type assertions to see if the output `io.Writer` is in fact also an `io.Seeker`, but Stdout can't seek, it's sort of broken.

I doubt writing to stdout is worthwihle?

jhford replied:
hmm, i could probably check if output == os.Stdout... I dunno.

jhford replied:
In the end, I decided that because these are extra checks, I'll instead try to seek, and only if the seek succeeds and the filesize isn't == 0 throw an error.

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+
+				if c.Bool("latest") {
+					if c.NArg() != 2 {
+						return cli.NewExitError("--latest requires two arguments", ErrBadUsage)
```
would be useful to state what arguments were passed, so it can be seen why it failed (logs might not show how the command was invoked, and then it is crystal clear what was called, and why it isn't allowed)

jhford replied:
You caught me being lazy :)

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+					err = client.DownloadLatest(c.Args().Get(0), c.Args().Get(1), output)
+				} else {
+					if c.NArg() != 3 {
+						return cli.NewExitError("three arguments required", 1)
```
same here

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+			Flags: []cli.Flag{
+				cli.StringFlag{
+					Name:   "input, i",
+					Usage:  "`FILENAME` to read as artifact.  Must be seekable",
```
I think all files represented by filenames should be seekable, so we should be safe to drop "Must be seekable".

jhford replied:
Named pipes have filenames in the filesystem but cannot have a seek operation done to them as far as I'm aware.

petemoore replied:
Ah, nice catch. Ok, let's leave it as is!

petemoore commented on cmd/artifact/main.go:
```diff
@@ -0,0 +1,341 @@
+					return cli.NewExitError(err, ErrInternal)
+				}
+				defer output.Close()
+				defer os.Remove(output.Name())
```
These should be in opposite order, or should be declared in a single defer block. Defer blocks are popped off the stack in the reverse order to how they were pushed on the stack.

jhford replied:
good catch

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+			}
+			w.Write(b)
+		}
+	}))
```
I really like this!

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+}
+
+func TestCLIRuns(t *testing.T) {
+	e, teardown := setup(t)
```
I really like how the setup function returns a teardown function - very neat.

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+		e.validate()
+	})
+
+}
```
Should we add a test that `--latest` really uses latest run by claiming/resolving a task twice e.g. resolving as `exception/worker-shutdown` and then `resolved` to get two task runs?

petemoore commented on cmd/artifact/test.sh:
```diff
@@ -0,0 +1,11 @@
+
+artifact upload $TASKID $RUNID public/artifact --input artifact
+artifact download $TASKID $RUNID public/artifact --output artifact-downloaded
+diff artifact artifact-downloaded
```
Should we delete this file? I think it isn't used, is it?

jhford replied:
yep!

petemoore commented on .gitignore:
```diff
@@ -12,3 +12,42 @@
+tags
+
+# End of https://www.gitignore.io/api/go,git,vim
+
```
might not be bad to add `vendor` here so any internal vendor directories don't get added

jhford replied:
so my understanding of vendoring is that we'd want to check in all the vendored code into our repository if we decide to vendor?  Is the thinking here that since we're not vendoring we shouldn't allow checking in of vendored code?

petemoore replied:
It is a fair point that we shouldn't decide on this until we adopt a vendoring strategy, you are right - so let's leave this as it is.

> so my understanding of vendoring is that we'd want to check in all the vendored code into our repository

Traditionally both models are commonly used - checked-in vendor directory, and not-checked-in vendor directory - it mostly depends on the strategy of the vendor tool being used. For example, sometimes only a manifest is checked in, and then some build tool checks out dependencies based on commit SHAs in external repos - other strategies include copying all source in-tree, so you don't need to rely on those external repos keeping those commits in them.

petemoore reviewed (commented):
We should also add either a `.travis.yml` or `.taskcluster.yml` file to this project to get integration testing.

Overall I really love the work you've done here - it is *really* nicely written - amazing it is your first go project!

petemoore commented on README.md:
```diff
@@ -1,2 +1,234 @@
+again for the upload
+
+- - -
+Generated by [godoc2ghmd](https://github.com/GandalfUK/godoc2ghmd)
```
nice idea to add auto-generated docs here - brilliant!

petemoore commented on body.go:
```diff
@@ -0,0 +1,71 @@
+	return nil
+}
+
+// Satisfy the io.Reader interface by reading from the associated file
```
nit: I'm pretty sure this won't pass `golint` validation as normally the description has to start with the function/method name.

jhford replied:
my experience is that golint only cares about that case for exported things, which body isn't.  It might be worthwhile writting anyway

petemoore replied:
OK, fair enough. :)

petemoore commented on body.go:
```diff
@@ -0,0 +1,71 @@
+	"io"
+)
+
+type body struct {
```
some explanation of `backingReader` and `limitReader` may be useful here to distinguish them

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,172 @@
+
+		if !bytes.Equal(bodyData, b[1024:]) {
+			t.Fatalf("Body data did not match")
+		}
```
I think this won't catch if it reads the first 1024 bytes, since the first and second 1024 bytes are the same. :-)

jhford replied:
it does now that it's random ;)

petemoore replied:
unless `USE_REPEATED_BYTES` env var is set, right?

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,172 @@
+
+		if !bytes.Equal(bodyData, b[512:1024+512]) {
+			t.Fatalf("Body data did not match")
+		}
```
same as before - the middle bytes are also indistinguishable from the bytes starting at the middle of the file

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,172 @@
+		if err != nil {
+			t.Fatal(err)
+		}
+	}
```
I think we shouldn't have repeated sections in the bytes, otherwise we can't be sure which section of a file a series of bytes represents, since it could be from any of the repetitions of that sequence. Maybe better to have non repetitions, or to generate a crypto-random stream of 2048 bytes?

jhford replied:
I used this particular pattern because it made finding errors easier, since as long as I read a certain chunksize, the beginning of each chunk/part would have the same sequence.

I'll change this and double check that the tests still work

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+	if !bytes.Equal(f1, f2) {
+		e.t.Error("File %s and %s unexpectedly differ", e.inputFilename, e.outputFilename)
+	}
+}
```
nice!

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+		Priority:      "lowest",
+		TaskGroupID:   taskGroupID,
+		WorkerType:    "my-workertype",
+	}
```
Are the scopes that are needed for these requests covered by the role [project:taskcluster:contributor](https://tools.taskcluster.net/auth/roles/project%3Ataskcluster%3Acontributor) already?

If not, let's create a role like `project:taskcluster:taskcluster-lib-api-go-tester` which has the required scopes for running the tests here, and then add `assume:project:taskcluster:taskclusster-lib-api-go-tester` to role `project:taskcluster:contributor` so that anyone wanting to contribute to this has an easy to get the test scopes (see [instructions](https://github.com/taskcluster/generic-worker/#acquire-taskcluster-credentials-for-running-tests)).

jhford replied:
it looks like it might.... if we use specific names, but I'd like to use descriptive names and not overwrite artifacts in the test.  I'll look into this as part of enabling travis or taskcluster on this repository.

petemoore commented on cmd/artifact/main_test.go:
```diff
@@ -0,0 +1,328 @@
+		e.validate()
+		run("download", e.taskID, name, "--latest", "--output", e.outputFilename)
+		e.validate()
+	})
```
This is a more concise version of the same:

```go
func TestCLIRuns(t *testing.T) {
	e, teardown := setup(t)
	defer teardown()

	// validateUploadOptions tests the downloaded file of an upload with the
	// given options is unaltered from the original file
	validateUploadOptions := func(name string, uploadOptions ...string) {
		t.Run(name, func(t *testing.T) {
			name := "public/" + name
			run("upload", e.taskID, e.runID, name, "--input", e.inputFilename, uploadOptions, uploadOptions...)
			run("download", e.taskID, e.runID, name, "--output", e.outputFilename)
			e.validate()
			run("download", e.taskID, name, "--latest", "--output", e.outputFilename)
			e.validate()
		})
	}

	validateUploadOptions("auto-identity") // no upload options
	validateUploadOptions("auto-gzip", "--gzip")
	validateUploadOptions("single-part-identity", "--single-part")
	validateUploadOptions("single-part-gzip", "--single-part", "--gzip")
	validateUploadOptions("multipart-identity", "--multi-part")
	validateUploadOptions("multipart-gzip", "--multi-part", "--gzip")
}
```

jhford replied:
yah, much nicer!  I did have to make a slight modification to make the compiler happy though:
```go
			upargs := []string{"upload", e.taskID, e.runID, name, "--input", e.inputFilename}
			upargs = append(upargs, uploadOptions...)
			run(t, upargs...)
```

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// Queue's blob artifacts.  Blob artifacts are the way that Taskcluster stores
+// and distributes the results of a task and replace the old "S3" type of
+// artifact for storing artifacts in S3.  These artifacts have stronger
+// authenticity and integrity guaruntees than the old type.
```
nit: old type -> former type

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+//
+// Blob artifacts can be between 1 byte and 5GB if uploaded as a single part
+// upload and between 1 byte and 5TB if uploaded as a multipart upload.  To
+// upload this type of artifact, the uploader must compute the artifacts sha256
```
artifact's

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// upload is complete.
+//
+// The queue ensures that the sha256 and size values are set as headers on the
+// artifacts in S3 so that downloaders can do verification
```
missing '.' at end of sentence - although I prefer the alternative ending " .... so that downloaded content can be verified."

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// The queue ensures that the sha256 and size values are set as headers on the
+// artifacts in S3 so that downloaders can do verification
+//
+// When downloading a blob artifact, the downloader must verify that the
```
When -> After
must -> should

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+//
+// The input and output parameters are various types of specialized io.Reader
+// and io.Writer types.  The minimum interface for use in the specific function
+// was chosen.  This library does not do any management of the input and output
```
I think the sentence "The minimum interface for use in the specific function was chosen." is superfluous and can be removed.

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// was chosen.  This library does not do any management of the input and output
+// objects.  They must be created outside of this library and any cleanup must
+// occur in calling code.  The most common output option is likely an
+// ioutil.TempFile() instance.
```
Let's put this last sentence in the method docs rather than the general docs.

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// io.Writer, but the output passed in does implement io.Seeker, this check is
+// also performed.  If the passed io.Writer really does not implement
+// io.Seeker, it is the responsibility of the caller to ensure it is refering
+// to an empty resource
```
Again, I think this is better served in the docs for the given functions, as at this point of reading the overview, the user isn't familiar with the API yet.

petemoore commented on docs.go:
```diff
@@ -0,0 +1,56 @@
+// calling code.  This could result in double gzip encoding if a gzip file is
+// passed into Upload() with the gzip argument set to true.  When this artifact
+// is downloaded with this library, the resulting output will be written as a
+// once encoded gzip file
```
Again, I think this paragraph should be in the docs for the upload function, not in the general package docs.

petemoore commented on log.go:
```diff
@@ -0,0 +1,38 @@
+// method.  See: https://golang.org/pkg/log/#Logger.SetFlags
+func SetLogFlags(f int) {
+	logger.SetFlags(f)
+}
```
It is a neat trick to just expose only wrapper functions that alter the logging options rather than exporting or exposing the entire underlying logger. I guess this is to prevent that someone logs to the internal logger? Would it be useful to have a function where you can pass in a custom `log.Logger` in case you have set one up for your entire application, and want your application dependencies to write to the same log with the same flags and prefix? This may save a few lines for users of the library that already have a custom log.Logger that they are using?

jhford committed:
better error messages

jhford committed:
resolve a nil pointer error

also improve logging in tests

jhford committed:
Fix a weird corner case with url.Error

jhford committed:
fix a small cli test

jhford committed:
use testdata directory

jhford committed:
fixup

jhford committed:
allow stdout as output

jhford committed:
Allow downloading from Queue URLs directly

jhford committed:
Better errors in the CLI and simplified error handling

Basically, urfave/cli just doesn't allow us to hook our own exit logic
into the places that I need to be able to.  It is detecting things
correctly, but I can't intercept it.  Because of that, I'll remove the
ErrBadUsage condition and just have a special ErrCorrupt, which I have
unit tests to show is set correctly when the error is a corruption issue

jhford committed:
use units library to parse KB instead of using hardcoded units

jhford committed:
mutlipart is one word without hyphen

jhford committed:
some review comments addressed

jhford commented:
@petemoore Mind taking another look?  If there's something totally missing, it might be because I did something really dumb yesterday and had to rebase a whole bunch of patches.

jhford committed:
Seem to have forgotten to commit a couple things

jhford committed:
linters

petemoore commented on body_test.go:
```diff
@@ -0,0 +1,189 @@
+	randbuf := make([]byte, 8)
+
+	for i := 0; i < 256; i++ {
+		// Using repeated bytes is nice for debugging purposes
```
Would that not make it harder, because you don't which of the repeated sections it was copied from? If they were random bytes, there is a good chance they could have only be copied from a single location from the original binary block, isn't there?

jhford replied:
it made it easier for me because I knew any number of bytes that was a multiple of 8 was going to have the same sequence.  Since I was trying to find off-by-one type things and using chunks which were sized as multiples of 8, I could spot the shift.  The bytes I chose even nicely printed into the whole screen and you could see a neat pattern.  It's probably served it's purpose and could be removed.... no strong feelings :)

petemoore reviewed (approved):
Looks awesome! Nice work! :-)

jhford commented:
Thanks!  And thanks for the helpful review feedback, I learned a bunch through it.
